### PR TITLE
[pylint] Added in custom PyLint Checker for Aliasing, with tests

### PR DESCRIPTION
**Pull request by l0lawrence**
Custom PyLint Checker for Aliasing, with basic testing #23324

### CODE REVIEW

File: scripts/pylint_custom_plugin/pylint_guidelines_checker.py [hunk 1]
```diff
@@ -1706,6 +1706,67 @@ def visit_functiondef(self, node):
...
 
 
+class CheckNoAliasGeneratedCode(BaseChecker):
+    __implements__ = IAstroidChecker
+
+    name = "check-alias"
+    priority = -1
+    msgs = {
+        "C4745": (
+            "Aliasing and exposing generated code."
```
Comment (iscai-msft) [inline scripts/pylint_custom_plugin/pylint_guidelines_checker.py:1716]: small nit (sorry if this goes against what i said earlier): I think saying "Exposing aliased generated code" is slightly better

**Review by iscai-msft — APPROVED** [review]
think this looks good! do you know how many SDKs have this issue?

one question to @johanste @lmazuel @annatisch we're focusing on generated code here since it's def the main scenario someone would change a model's name with aliasing. Do you guys want to make it more generic?

**Review by catalinaperalta — APPROVED** [review]
LGTM! Left one question

File: scripts/pylint_custom_plugin/pylint_guidelines_checker.py [hunk 2]
```diff
@@ -1706,6 +1706,67 @@ def visit_functiondef(self, node):
...
+                        for name in nod.names:
+                            if name[1] != None:
+                                aliased.append(name[1])
+
+                    if isinstance(nod, astroid.Assign):
+                        for models in nod.assigned_stmts():
+                            for model_name in models.elts:
+                                if model_name.value in aliased:
+                                    self.add_message(
+                                        msgid="aliasing-generated-code", node=model_name, confidence=None
```
Comment (kristapratico) [inline scripts/pylint_custom_plugin/pylint_guidelines_checker.py:1762]: just curious, does `model_name` contain a `lineno`? If we put a pylint: disable=aliasing-generated-code on the offending line, does it correctly ignore the error?

File: scripts/pylint_custom_plugin/tests/test_pylint_custom_plugins.py [hunk 3]
```diff
@@ -2568,4 +2568,73 @@ def __init__(self):
...
+        with self.assertNoMessages():
+            self.checker.visit_module(module_node)
+
+    def test_catches_incorrect_alias_code(self):
+        import_one = astroid.extract_node(
+            'import Something'
+           
+        )
+        import_two =  astroid.extract_node(
+            'import Something2 as SomethingTwo'
```
Comment (kristapratico) [inline scripts/pylint_custom_plugin/tests/test_pylint_custom_plugins.py:2593]: I think we should have a `from something import something1 as something2` as a scenario in the tests

File: scripts/pylint_custom_plugin/pylint_guidelines_checker.py [hunk 4]
```diff
@@ -1706,6 +1706,67 @@ def visit_functiondef(self, node):
...
+                aliased = []
+              
+                for nod in node.body:
+                    if isinstance(nod, astroid.ImportFrom) or isinstance(nod, astroid.Import):
+                        # If the model has been aliased
+                        for name in nod.names:
+                            if name[1] != None:
+                                aliased.append(name[1])
+
+                    if isinstance(nod, astroid.Assign):
```
[the remark this answers is not in the document]
Comment (l0lawrence) [reply scripts/pylint_custom_plugin/pylint_guidelines_checker.py:1757]: just in case - pull it out of the loop (offline)

**Comment by l0lawrence** [comment]
>

File: scripts/pylint_custom_plugin/tests/test_files/__init__.py [hunk 5]
```diff
@@ -0,0 +1,7 @@
+from something import Something
+from something2 import something2 as somethingTwo
+
+__all__ = (
+    Something,
+    somethingTwo, #pylint: disable=aliasing-generated-code
+)
```
Comment (catalinaperalta) [inline scripts/pylint_custom_plugin/tests/test_files/__init__.py:7]: Nit: missing newline

Comment (catalinaperalta) [reply scripts/pylint_custom_plugin/tests/test_files/__init__.py:7]: Not saying we should do this now, but adding more test files for different test cases could be cool. Like to check that the error message includes the proper line, etc.

**Review by catalinaperalta — APPROVED** [review]

File: scripts/pylint_custom_plugin/tests/test_files/__init__.py (hunk 5 above)
Comment (l0lawrence) [reply scripts/pylint_custom_plugin/tests/test_files/__init__.py]: I will make an issue for it

File: scripts/pylint_custom_plugin/pylint_guidelines_checker.py [hunk 6]
```diff
@@ -1706,6 +1706,67 @@ def visit_functiondef(self, node):
...
+class CheckNoAliasGeneratedCode(BaseChecker):
+    __implements__ = IAstroidChecker
+
+    name = "check-alias"
+    priority = -1
+    msgs = {
+        "C4745": (
+            "Exposing aliased generated code."
+            "This messes up sphinx, intellisense, and apiview, so please modify the name of the generated code through"
+            " the swagger / directives, or code customizations",
```
[the remark this answers is not in the document]
Comment (l0lawrence) [reply scripts/pylint_custom_plugin/pylint_guidelines_checker.py:1718]: https://github.com/Azure/autorest/blob/main/docs/generate/built-in-directives.md (from Izzy)

**Review by catalinaperalta — APPROVED** [review]
Great catch on updating the links in the README!

**Pull request merged**